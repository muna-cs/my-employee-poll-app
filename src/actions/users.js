export const RECEIVE_USERS = "RECEIVE_USERS";
export const SAVE_ANSWER="SAVE_ANSWER";
export const SAVE_QUESTION="SAVE_QUESTION";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function saveAnswer(authedUser, qid, answer) {
  return {
    type: SAVE_ANSWER,
    authedUser,
    qid,
    answer,
  };

};

export function saveQuestion(question) {
  return {
    type: SAVE_QUESTION,
    question
  };
};