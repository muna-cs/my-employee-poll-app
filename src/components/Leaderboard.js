import { Container, Table ,Row,Image} from "react-bootstrap";
import { connect } from "react-redux";
import React from 'react'
const Leaderboard= (props) =>{

return(
        

    <Container className=" bg-light p-5">

        <Row className="text-center mt-3">
       <h2 className="text-primary">Leaderboard</h2>
       <hr/>
        </Row>
        <Row>
        <Table data-testid='UserList'  striped bordered hover>
      <thead>
        <tr>
          <th> Name</th>
          <th>Answered Questions</th>
          <th>Created Questions</th>
        </tr>
      </thead>
      <tbody>
        
          {props.UsersInfo.map((usr)=>
          <tr key={usr.id}>
            <td ><Image src={usr.avatarURL} width='50' alt="Logo" /> <span className="text-dark h5">{usr.name}</span></td>
            <td>{Object.keys(usr.answers).length}</td>
            <td>{usr.questions.length}</td>
            </tr>
            )}
       
        </tbody>
        </Table></Row>
    </Container>
);

}
const mapStateToProps = ({authedUser, users }) => ({
    authedUser:users[authedUser],
    UsersInfo: Object.values(users).sort(
      (a, b) => (Object.keys(b.answers).length + Object.keys(b.questions).length) -
       (Object.keys(a.answers).length + Object.keys(a.questions).length)
    ),

})
export default connect(mapStateToProps)(Leaderboard);