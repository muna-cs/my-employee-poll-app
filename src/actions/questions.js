import {saveNewQuestion, saveNewQuestionAnswer }from '../utils/api'
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { saveAnswer , saveQuestion } from './users';
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER = "ADD_ANSWER";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addAnswer(authedUser,qid,answer) {
  return {
    type: ADD_ANSWER,
    authedUser, qid, answer
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question, 
  };
}

export function handleAddAnswer(authedUser,qid,answer) {
  return (dispatch) => {
    dispatch(addAnswer(authedUser, qid, answer));
    dispatch(saveAnswer(authedUser, qid, answer));
    return saveNewQuestionAnswer({authedUser:authedUser, qid, answer})
    .catch((e) => {
      console.warn("Error in Add Answer: ", e);
      dispatch(addAnswer(authedUser, qid, answer));
      dispatch(saveAnswer(authedUser, qid, answer));
    });
       
      
};
}


export function handleAddQuestion(authedUser,optionOneText,optionTwoText) {
  return (dispatch) => {
    dispatch(showLoading());
    return saveNewQuestion({ optionOneText, optionTwoText,author:authedUser})
       .then((q) => {
       dispatch(addQuestion(q));
       dispatch(saveQuestion(q));
        }).then(() => dispatch(hideLoading()));
};
}