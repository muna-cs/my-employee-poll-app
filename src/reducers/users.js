import { RECEIVE_USERS ,SAVE_ANSWER, SAVE_QUESTION } from "../actions/users";

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
      case SAVE_ANSWER:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
            
          },
        },
      };
      case SAVE_QUESTION:
      return {
        ...state,
        [action.question.author]: {
          ...state[action.question.author],
          questions: [action.question.id].concat(state[action.question.author].questions),
        },
      };
    default:
      return state;
  }
}