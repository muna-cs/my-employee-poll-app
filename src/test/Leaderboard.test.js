import * as React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter as Router } from 'react-router-dom';
import { Provider } from "react-redux";
import { configureStore } from '@reduxjs/toolkit';
import authedUser from '../reducers/authedUser';
import users from '../reducers/users';
import questions from '../reducers/questions';
import logger from '../middleware/logger';
import thunk from 'redux-thunk';
import Leaderboard from '../components/Leaderboard';

export const store = configureStore({
  reducer: {
    authedUser,
    users,
    questions
  },
  middleware: [thunk, logger]
});
var component =  render(
    <Provider store={store}>
<Router>
<Leaderboard/>
   </Router>
   </Provider>
    );
describe('Testing Leaderboard', () => {
   
    it('it will create create snapshot', () => {
        var component =  render(
            <Provider store={store}>
        <Router>
        <Leaderboard/>
           </Router>
           </Provider>
            );
            expect(component).toMatchSnapshot();
        });

        it('it will verify if expected user list is exist', () => {
            var component =  render(
                <Provider store={store}>
            <Router>
            <Leaderboard/>
               </Router>
               </Provider>
                );
            var UserList = component.getByTestId('UserList')
            expect(UserList).toBeInTheDocument();
           
            
    
             });

    });
