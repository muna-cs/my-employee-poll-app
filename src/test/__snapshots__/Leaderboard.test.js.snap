// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Leaderboard it will create create snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=" bg-light p-5 container"
      >
        <div
          class="text-center mt-3 row"
        >
          <h2
            class="text-primary"
          >
            Leaderboard
          </h2>
          <hr />
        </div>
        <div
          class="row"
        >
          <table
            class="table table-striped table-bordered table-hover"
            data-testid="UserList"
          >
            <thead>
              <tr>
                <th>
                   Name
                </th>
                <th>
                  Answered Questions
                </th>
                <th>
                  Created Questions
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </div>
    <div>
      <div
        class=" bg-light p-5 container"
      >
        <div
          class="text-center mt-3 row"
        >
          <h2
            class="text-primary"
          >
            Leaderboard
          </h2>
          <hr />
        </div>
        <div
          class="row"
        >
          <table
            class="table table-striped table-bordered table-hover"
            data-testid="UserList"
          >
            <thead>
              <tr>
                <th>
                   Name
                </th>
                <th>
                  Answered Questions
                </th>
                <th>
                  Created Questions
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=" bg-light p-5 container"
    >
      <div
        class="text-center mt-3 row"
      >
        <h2
          class="text-primary"
        >
          Leaderboard
        </h2>
        <hr />
      </div>
      <div
        class="row"
      >
        <table
          class="table table-striped table-bordered table-hover"
          data-testid="UserList"
        >
          <thead>
            <tr>
              <th>
                 Name
              </th>
              <th>
                Answered Questions
              </th>
              <th>
                Created Questions
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
